org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/InvisiblePreferenceChangeListener.java

```diff
@@ -0,0 +1,54 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.jdt.ls.core.internal.managers;
+
+import java.util.Objects;
+
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.ls.core.internal.ActionableNotification;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.ProjectUtils;
+import org.eclipse.jdt.ls.core.internal.preferences.IPreferencesChangeListener;
+import org.eclipse.jdt.ls.core.internal.preferences.Preferences;
+import org.eclipse.lsp4j.MessageType;
+
+public class InvisiblePreferenceChangeListener implements IPreferencesChangeListener {
+
+	@Override
+	public void preferencesChange(Preferences oldPreferences, Preferences newPreferences) {
+		boolean projectOutputPathChanged = !Objects.equals(oldPreferences.getInvisibleProjectOutputPath(), newPreferences.getInvisibleProjectOutputPath());
+		if (projectOutputPathChanged) {
+			for (IJavaProject javaProject : ProjectUtils.getJavaProjects()) {
+				IProject project = javaProject.getProject();
+				if (ProjectUtils.isVisibleProject(project)) {
+					continue;
+				}
+				if (project.equals(ProjectsManager.getDefaultProject())) {
+					continue;
+				}
+				try {
+					ProjectUtils.setInvisibleProjectOutputPath(javaProject, newPreferences.getInvisibleProjectOutputPath(), new NullProgressMonitor());
+				} catch (CoreException e) {
+					ActionableNotification invalidOutputPathNotification = new ActionableNotification().withSeverity(MessageType.Error)
+							.withMessage(e.getMessage());
+					JavaLanguageServerPlugin.getProjectsManager().getConnection().sendActionableNotification(invalidOutputPathNotification);
```

Comment:
need to check preferencesManager.getClientPreferences().isActionableRuntimeNotificationSupport(), else send regular notification

Reply:
Thanks for pointing out! Since there is actually no actions here, I changed to the regular notification directly.